compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -570,7 +570,40 @@ pub struct DerivedObligationCause<'tcx> {
 }
 
 #[derive(Clone, Debug, TypeVisitable, Lift)]
-pub enum SelectionError<'tcx> {
+pub struct SelectionError<'tcx> {
+    /// The boxed kind. `None` represents `Unimplemented`, the only case present
+    /// in successful compilations and therefore the most common one.
+    kind: Option<Box<SelectionErrorKind<'tcx>>>,
+}
+
+impl<'tcx> SelectionError<'tcx> {
+    pub fn new(kind: SelectionErrorKind<'tcx>) -> Self {
+        let kind =
+            if let SelectionErrorKind::Unimplemented = kind { None } else { Some(Box::new(kind)) };
+        Self { kind }
```

Comment:
```suggestion
        SelectionError { kind }
```
`Self` in non-generic contexts is bad for code searchability and readability.